src/content_mapblock.cpp

```diff
@@ -525,1633 +371,955 @@ static TileSpec getSpecialTile(const ContentFeatures &f,
+	static const bool check_nb_all        [18] = { 1,1,1,1,1,1, 1,1,1,1, 1,1,1,1, 1,1,1,1 };
+	const bool *check_nb = check_nb_all;
+
+	// neighbours checks for frames visibility
+	if (H_merge || V_merge) {
+		if (!H_merge)
+			check_nb = check_nb_vertical; // vertical-only merge
+		if (!V_merge)
+			check_nb = check_nb_horizontal; // horizontal-only merge
+		content_t current = n.getContent();
+		for (int i = 0; i < 18; i++) {
+			if (!check_nb[i])
+				continue;
+			v3s16 n2p = blockpos_nodes + p + g_26dirs[i];
+			MapNode n2 = data->m_vmanip.getNodeNoEx(n2p);
+			content_t n2c = n2.getContent();
+			if (n2c == current || n2c == CONTENT_IGNORE)
+				nb[i] = 1;
+		}
+	}
 
-			for (s32 i = 0; i < 4; i++)
-			{
-				if(dir == v3s16(1,0,0))
-					vertices[i].Pos.rotateXZBy(0);
-				if(dir == v3s16(-1,0,0))
-					vertices[i].Pos.rotateXZBy(180);
-				if(dir == v3s16(0,0,1))
-					vertices[i].Pos.rotateXZBy(90);
-				if(dir == v3s16(0,0,-1))
-					vertices[i].Pos.rotateXZBy(-90);
-				if(dir == v3s16(0,-1,0))
-					vertices[i].Pos.rotateXZBy(45);
-				if(dir == v3s16(0,1,0))
-					vertices[i].Pos.rotateXZBy(-45);
-
-				if (data->m_smooth_lighting)
-					vertices[i].Color = blendLight(frame, vertices[i].Pos, tile.color);
-				vertices[i].Pos += intToFloat(p, BS);
-			}
+	// edge visibility
 
-			u16 indices[] = {0,1,2,2,3,0};
-			// Add to mesh collector
-			collector.append(tile, vertices, 4, indices, 6);
-		break;}
-		case NDT_SIGNLIKE:
-		{
-			TileSpec tile = getNodeTileN(n, p, 0, data);
-			tile.material_flags &= ~MATERIAL_FLAG_BACKFACE_CULLING;
-			tile.material_flags |= MATERIAL_FLAG_CRACK_OVERLAY;
-
-			u16 l = getInteriorLight(n, 0, nodedef);
-			video::SColor c = encode_light_and_color(l, tile.color,
-				f.light_source);
-
-			float d = (float)BS/16;
-			float s = BS/2*f.visual_scale;
-			// Wall at X+ of node
-			video::S3DVertex vertices[4] =
-			{
-				video::S3DVertex(BS/2-d,  s,  s, 0,0,0, c, 0,0),
-				video::S3DVertex(BS/2-d,  s, -s, 0,0,0, c, 1,0),
-				video::S3DVertex(BS/2-d, -s, -s, 0,0,0, c, 1,1),
-				video::S3DVertex(BS/2-d, -s,  s, 0,0,0, c, 0,1),
-			};
+	static const u8 nb_triplet[12*3] = {
+		1,2, 7,  1,5, 6,  4,2,15,  4,5,14,
+		2,0,11,  2,3,13,  5,0,10,  5,3,12,
+		0,1, 8,  0,4,16,  3,4,17,  3,1, 9
+	};
 
-			v3s16 dir = n.getWallMountedDir(nodedef);
-
-			for (s32 i = 0; i < 4; i++)
-			{
-				if(dir == v3s16(1,0,0))
-					vertices[i].Pos.rotateXZBy(0);
-				if(dir == v3s16(-1,0,0))
-					vertices[i].Pos.rotateXZBy(180);
-				if(dir == v3s16(0,0,1))
-					vertices[i].Pos.rotateXZBy(90);
-				if(dir == v3s16(0,0,-1))
-					vertices[i].Pos.rotateXZBy(-90);
-				if(dir == v3s16(0,-1,0))
-					vertices[i].Pos.rotateXYBy(-90);
-				if(dir == v3s16(0,1,0))
-					vertices[i].Pos.rotateXYBy(90);
-
-				if (data->m_smooth_lighting)
-					vertices[i].Color = blendLight(frame, vertices[i].Pos, tile.color);
-				vertices[i].Pos += intToFloat(p, BS);
-			}
+	tile = tiles[1];
+	for (int edge = 0; edge < 12; edge++) {
```

Comment:
why? is there any rule for that? IMO `int` is cleaner, and the size does not matter.